### PR TITLE
526 disability increase array field refactor

### PR DESCRIPTION
**Pull request by cehsu**
These changes update the review mode logic so that it is based on data availability.

screen shot 2018-04-25 at 11 55 46 am

This includes:
 - prefilled items are shown in a condensed view
 - invalid or empty prefilled items are shown in an unfurled view
 - empty items are no longer able to be added (in cases where schemas have no required fields)

Therefore:

1. valid items are now collapsed by default (this specifically affects the last item, which was previously displayed in edit mode with the white background "add another item" styling by default)
image

2. any invalid/incomplete items (including non-last items) are now displayed in edit mode
image

3. prefilled empty items with schemas that do not contain any required fields are displayed unfurled. This scenario is encountered if a user "adds another" saves the form, and returns later to continue filling the form out.
screen shot 2018-05-30 at 5 56 57 pm

4. empty items can no longer be added (in cases where schemas do not contain required fields, and no validation logic already prevents this)
This helps to determine whether empty items with schemas that do not contain any required fields are displayed properly. This scenario is encountered if a user "adds another" saves the form, and returns later to continue filling the form out.
screen shot 2018-05-31 at 10 02 25 am

### CODE REVIEW

**Comment by cehsu** [comment]
@cvalarida @markgreenburg is this consistent with what we discussed yesterday? Looping in @jbalboni for visibility.

File: src/js/common/schemaform/fields/ArrayField.jsx
Comment (cvalarida) [inline src/js/common/schemaform/fields/ArrayField.jsx]: Can you remind me again what review mode does?

Comment (jbalboni) [reply src/js/common/schemaform/fields/ArrayField.jsx]: I'm also not sure I understand why this change is being made.

Comment (cehsu) [reply src/js/common/schemaform/fields/ArrayField.jsx]: The 526 form will display the review form of the array field if data is available (prefilled), rather than the standard form view initially. 
screen shot 2018-04-25 at 11 55 46 am
If we want to hold off on this for the rest of the forms for now, perhaps we should create a 526 specific array field?

Comment (jbalboni) [reply src/js/common/schemaform/fields/ArrayField.jsx]: Yeah, I think this will have some negative effects on other forms, so we should come up with a different way of achieving it.

Off the top of my head, we might be able to add an option to ArrayField that starts with all items in the read only state if some condition is met.

Comment (cehsu) [reply src/js/common/schemaform/fields/ArrayField.jsx]: Sounds good, thanks @jbalboni, we will need to pass the formConfig via routes in order to have access to that form config option, I'll push up those changes for your consideration.

Comment (cehsu) [reply src/js/common/schemaform/fields/ArrayField.jsx]: We currently have it set up to specify this in uiSchema, but array field needs to be passed the formConfig from routes rather than from form in order to access that information. I'll make the changes to provide that access, and then update the reviewMode logic to also require that config setting.

File: src/js/common/schemaform/fields/ArrayField.jsx
Comment (cehsu) [inline src/js/common/schemaform/fields/ArrayField.jsx]: @jbalboni Arrrayfield already had the necessary access to uiSchema, so have simply included editingState here.

**Comment by cvalarida** [comment]
When there are no items in the array and `reviewMode` is true, this is what happens:

image

This is because the `formData` passed to `ArrayField` looks like
```javascript
[{ dateRange: {} }]
```
I forget why this happens, to be honest, but it was for good reason, I'm sure. _(Waves hands.)_

Line 197 then returns true:
```javascript
const isReviewMode = formData && formData.length && editingState;
```

So we're passing a basically-empty-but-not-really-empty object to the `viewField` for the first service period.

**Comment by jbalboni** [comment]
I don't think the right change here to change how the review mode flag works. I think the thing we want to tweak is how the last item is handled when data exists for the array.

So I would make a flag in `ui:options` called something like `showLastItemInViewMode` and use that to tweak the array field component logic where `isLast` is used to show the last item in view mode.

**Comment by jbalboni** [comment]
`reviewMode` is the prop we use to determine what set of widgets to display, the review ones or the normal ones. I'd rather not rename that, since it lines up with the review page wording.

File: src/applications/common/schemaform/fields/ArrayField.jsx
Comment (cvalarida) [inline src/applications/common/schemaform/fields/ArrayField.jsx]: For some reason, I'm having a hard time follow this added chunk. @cehsu Can you explain what this doing?

**Review by cvalarida — COMMENTED** [review]
Hmm...I found a bug. Steps to reproduce:
1) Go to (military service history page)[http://localhost:3001/disability-benefits/526/apply-for-increase/review-veteran-details/military-service-history]
    - The page is pre-filled with a single service period
    - It has a validation error, so it starts in edit mode like it should
2) Enter the start date (which is required but is currently blank)
    - It immediately goes to the view mode as soon as I enter the year
3) Hit the edit button
    - I get sent to the next page (payment information)

**Comment by cvalarida** [comment]
Looks like Jenkins passed, which means none of the tests caught the bug I found. We should fix that.

I think this is a better approach than re-using the `reviewMode`, so I'd say it's heading in the right direction! Just needs some more work is all.

File: src/applications/common/schemaform/fields/ArrayField.jsx
Comment (cehsu) [reply src/applications/common/schemaform/fields/ArrayField.jsx]: This is checking that the last time (the item this change effects), is valid, to determine whether the showlastItemInViewMode configuration should be applied or not (if it is not valid).

Comment (cehsu) [reply src/applications/common/schemaform/fields/ArrayField.jsx]: Would it make sense to add some comments here to clarify?

File: src/applications/common/schemaform/fields/ArrayField.jsx
Comment (cehsu) [inline src/applications/common/schemaform/fields/ArrayField.jsx]: @cvalarida this change fixes the bug (view changing once user input is valid) you found.

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cehsu) [inline test/common/schemaform/fields/ArrayField.unit.spec.jsx]: @cvalarida this test has been set up for the bug you found.

File: src/applications/disability-benefits/526EZ/config/form.js
Comment (cehsu) [inline src/applications/disability-benefits/526EZ/config/form.js]: This requires that dateRange has a 'from' property to be valid.

**Comment by cehsu** [comment]
Sure thing @cvalarida, I've set up a ticket based on your comment above and connected to this PR.

File: src/applications/disability-benefits/526EZ/config/form.js
Comment (cvalarida) [reply src/applications/disability-benefits/526EZ/config/form.js]: Seems like this should be set in the schema, but this is probably okay. :man_shrugging:

File: src/applications/disability-benefits/526EZ/tests/schema/initialData.js
Comment (cvalarida) [inline src/applications/disability-benefits/526EZ/tests/schema/initialData.js]: Can we put these back in for ease of testing later?

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cvalarida) [inline test/common/schemaform/fields/ArrayField.unit.spec.jsx]: Is there a problem with using `errorSchema` here?

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cvalarida) [inline test/common/schemaform/fields/ArrayField.unit.spec.jsx]: Does calling it `review mode` here make sense? Seems like that could be confusing. What do you think of `view mode`?

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cvalarida) [inline test/common/schemaform/fields/ArrayField.unit.spec.jsx]: Might be worth making `viewField` a dummy component and also checking that it renders.

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cvalarida) [inline test/common/schemaform/fields/ArrayField.unit.spec.jsx]: Instead of `when specified and has no data`, would it be clearer to say `when showLastItemInViewMode is true and no data is given`?

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cvalarida) [inline test/common/schemaform/fields/ArrayField.unit.spec.jsx]: Same as above but with `invalid data is given`.

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cvalarida) [reply test/common/schemaform/fields/ArrayField.unit.spec.jsx]: What do you think of the test description text `should not exit edit mode prematurely when validation errors are resolved`? It took me a couple tries to figure out what this test should be doing.

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cvalarida) [inline test/common/schemaform/fields/ArrayField.unit.spec.jsx]: `thing` isn't a property of the array items. It will probably be more comprehensive if we make `field` required and then fill it out. It should start in the edit mode, and when we fill it out, it should stay in edit mode.

File: src/applications/common/schemaform/fields/ArrayField.jsx
Comment (markgreenburg) [reply src/applications/common/schemaform/fields/ArrayField.jsx]: Does this cover the case where hitting 'edit' under any circumstance advances to the next page? Cause I ran into that. Basically -> use existing, valid initialData -> hit 'Edit' next to the first (and only) item -> advances to payments page.

**Comment by markgreenburg** [comment]
Here's some feedback based solely on me playing around with this PR locally:
Here's some stuff that happened when I used a servicePeriod that I believe to be valid:
```{
      serviceBranch: 'Army',
      dateRange: {
        from: '1990-02-02',
        to: '2010-03-04'
      }
    }
```

1. The 'Edit' button (next to the service period) action seems inconsistent. I haven't been able to dive into why this might be happening yet.
  - Sometimes nothing happens
  - Sometimes I get advanced to the next page in the form flow (payments)
2. The 'Add another service period' button automatically adds a blank service period in 'view' mode (and clicking on the edit button yields the same results as point 1. above). This is probably meant to automatically open in 'edit' mode instead, I'd think

service history

File: src/applications/disability-benefits/526EZ/config/form.js
Comment (markgreenburg) [reply src/applications/disability-benefits/526EZ/config/form.js]: Schema sets both `from` and `to` as required, which matches `submit` swagger so should be used in the config. Also, just as FYI, I'm working on some changes that restrict the `serviceBranch` to an `enum` list per EVSS. But that's also getting added to the schema, so shouldn't require anything else here as long as the schema's version of the definition is used.

File: src/applications/common/schemaform/fields/ArrayField.jsx
Comment (cehsu) [reply src/applications/common/schemaform/fields/ArrayField.jsx]: Good catch, will update.

File: src/applications/disability-benefits/526EZ/config/form.js
Comment (cehsu) [reply src/applications/disability-benefits/526EZ/config/form.js]: Thanks @markgreenburg will update using the schema.

File: src/applications/disability-benefits/526EZ/tests/schema/initialData.js
Comment (cehsu) [reply src/applications/disability-benefits/526EZ/tests/schema/initialData.js]: Definitely.

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cehsu) [reply test/common/schemaform/fields/ArrayField.unit.spec.jsx]: That's fine too, review mode is used in the field itself, also, this is sort of the array equivalent of the review card field now, so that was probably why I used "review" here.

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cehsu) [reply test/common/schemaform/fields/ArrayField.unit.spec.jsx]: It is declared in the upper scope.

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cehsu) [reply test/common/schemaform/fields/ArrayField.unit.spec.jsx]: Yeah that's a better description, will update.

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cehsu) [reply test/common/schemaform/fields/ArrayField.unit.spec.jsx]: True, will update similarly.

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cehsu) [reply test/common/schemaform/fields/ArrayField.unit.spec.jsx]: How about `should not render review mode until valid data submitted` to clarify and try to keep the message short.

File: src/applications/common/schemaform/fields/ArrayField.jsx
Comment (cehsu) [reply src/applications/common/schemaform/fields/ArrayField.jsx]: Good eye, "New" can probably be replaced by "Additional" as it is applied to items only if they are the last of a list with a length longer than 1.
The update button is typically not available for the last item, but we can use local state to determine whether we should provide that (if it is accessed via the edit button rather than add another).
Remove shouldn't be available if there's only one item left, so that needs to be `removed` as well.

**Comment by cehsu** [comment]
There are two updates to the requirements here:

1. @goldenmeanie has clarified that the default behavior for the array field should now be to collapse any prefilled items, and therefore not require any configuration in the form config (previously achieved using `reviewMode` and was planned to be replaced by `showLastItemInViewMode`).
2. @markgreenburg has also explained that support for incomplete prefilled items is required.

Therefore:

1. valid items are now collapsed by default (this specifically affects the last item, which was previously displayed in edit mode with the add item styling by default)
image

2. any invalid items (including non-last items) are now displayed in edit mode
image

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cehsu) [reply test/common/schemaform/fields/ArrayField.unit.spec.jsx]: Good point, test updated.

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cehsu) [reply test/common/schemaform/fields/ArrayField.unit.spec.jsx]: Great idea, added.

File: src/applications/common/schemaform/fields/ArrayField.jsx
Comment (jbalboni) [inline src/applications/common/schemaform/fields/ArrayField.jsx]: What scenario did you hit that requires the `isEmpty` check? It seems like it would only be necessary in cases where an array had a `minItems` requirement, but none of the fields in the array items were required, which I don't believe we do anywhere (because it'd lead to bad data).

File: src/applications/common/schemaform/fields/ArrayField.jsx
Comment (jbalboni) [inline src/applications/common/schemaform/fields/ArrayField.jsx]: I think we might benefit from tweaking this state name and possible values with these changes. `true`, `false`, and `'adding'` seems confusing. Maybe we should call it `itemMode` and use `viewing`, `editing`, and `adding`?

**Review by jbalboni — COMMENTED** [review]
Left a couple comments. Please make sure to test these changes on our other forms, both on regular pages and review pages.

File: src/applications/common/schemaform/fields/ArrayField.jsx
Comment (cvalarida) [reply src/applications/common/schemaform/fields/ArrayField.jsx]: Yeah, thinking through this, I can't think of a time where this will do what we want it to. If we have an array of empty items, it should be fine unless the objects have required fields. Alternatively, if we had an empty array, that should be fine unless there's a `minItems` (in which case, the `errorSchema` won't be valid).

This here would catch if `minItems > 0` and those items were all blank and those items had no required fields.

File: src/applications/disability-benefits/526EZ/config/form.js
Comment (cvalarida) [inline src/applications/disability-benefits/526EZ/config/form.js]: Where is this being used?

File: src/applications/common/schemaform/fields/ArrayField.jsx
Comment (cehsu) [reply src/applications/common/schemaform/fields/ArrayField.jsx]: Thanks @jbalboni, that is much more expressive, updated.

File: src/applications/common/schemaform/fields/ArrayField.jsx
Comment (cehsu) [reply src/applications/common/schemaform/fields/ArrayField.jsx]: Thanks @cvalarida and @jbalboni, good point, `isEmpty` removed.

File: src/applications/disability-benefits/526EZ/config/form.js
Comment (cehsu) [reply src/applications/disability-benefits/526EZ/config/form.js]: This was actually changed in master.

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cvalarida) [inline test/common/schemaform/fields/ArrayField.unit.spec.jsx]: The new `expect` looks good, but this one confuses me, and I find that concerning. In this old test, this is my understanding:
1) We have two items in the array
2) The first item is rendered as not editing (`viewField`)
3) The second item is rendered as editing (because it's the last item in the array)
4) When we hit the add button, it puts the second item into a view state (not editing) and adds a third
5) This newly-added third item should be rendered in the edit state

So...why did we expect the edit state of the third item to be false? @jbalboni?

**Review by cvalarida — COMMENTED** [review]
It looks good to me, but I'd like to get some clarification on some things.

Can you outline what the expected behavior (acceptance criteria) is when the following cases are true?
- No data is pre-filled
- A single item is pre-filled, no validation errors
- Multiple items are pre-filled, no validation errors
- A single item is pre-filled with validation error
- Multiple items are pre-filled with validation error on the last item
- Multiple items are pre-filled with validation error on not the last item

File: src/applications/common/schemaform/fields/ArrayField.jsx
Comment (cehsu) [reply src/applications/common/schemaform/fields/ArrayField.jsx]: Good point @cvalarida, that is a great idea.

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cehsu) [reply test/common/schemaform/fields/ArrayField.unit.spec.jsx]: One thing that might be confusing here is that this no longer called `editing` but `itemModes`, and the new item's `itemMode` is now being tested as being `adding`, in case that helps to clarify what the change is here.

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cehsu) [reply test/common/schemaform/fields/ArrayField.unit.spec.jsx]: Could you please clarify why we would want to test whether the invalid prefilled item is the last or not?

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cehsu) [reply test/common/schemaform/fields/ArrayField.unit.spec.jsx]: Good idea, that's much clearer.

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cehsu) [reply test/common/schemaform/fields/ArrayField.unit.spec.jsx]: Interesting, form data is not being provided here, since it is testing an empty form. Would providing the formData prop improve the readability of this test?

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cehsu) [reply test/common/schemaform/fields/ArrayField.unit.spec.jsx]: Yes more description would be helpful here, perhaps, should render an empty form?

File: test/common/schemaform/fields/ArrayField.unit.spec.jsx
Comment (cehsu) [reply test/common/schemaform/fields/ArrayField.unit.spec.jsx]: Good point, I'll revert this to the original: `{ __errors: ['Test error'] };`

File: src/applications/common/schemaform/fields/ArrayField.jsx
Comment (cehsu) [inline src/applications/common/schemaform/fields/ArrayField.jsx]: This helps to determine whether empty items with schemas that do not contain any required fields are displayed properly. This scenario is encountered if a user "adds another" saves the form, and returns later to continue filling the form out.
screen shot 2018-05-30 at 5 56 57 pm

File: src/applications/common/schemaform/fields/ArrayField.jsx
Comment (cehsu) [inline src/applications/common/schemaform/fields/ArrayField.jsx]: This disables the "add another" button when no information has been provided, and prevents the creation of empty items (particularly when no fields are required by the schema/no validation prevents this from occuring).

**Comment by cehsu** [comment]
Thanks @jbalboni @cvalarida @markgreenburg for your comments which have been addressed. Thanks for taking another look when you can and kindly refer to the updated description for a summary of the changes. Thanks for any additional feedback as well.

File: src/applications/common/schemaform/fields/ArrayField.jsx
Comment (jbalboni) [reply src/applications/common/schemaform/fields/ArrayField.jsx]: Ok, that's fine for now. Can we make a ticket to go through the forms and update these cases to have at least one required field? I think it's a bad practice to have array fields where there are no required fields in the items. It encourages bad data being sent over. Once we do that, I think we could simplify how this works.

Comment (cehsu) [reply src/applications/common/schemaform/fields/ArrayField.jsx]: Thanks @jbalboni, issue created https://github.com/department-of-veterans-affairs/vets.gov-team/issues/10732.

File: src/applications/common/schemaform/fields/ArrayField.jsx
Comment (jbalboni) [inline src/applications/common/schemaform/fields/ArrayField.jsx]: Normally we use the regular function syntax for these functions defined outside of React components. Curious if you had a reason for defining them as arrow functions.

**Review by jbalboni — APPROVED** [review]

File: src/applications/common/schemaform/fields/ArrayField.jsx
Comment (cehsu) [reply src/applications/common/schemaform/fields/ArrayField.jsx]: Thanks @jbalboni, changed.

File: src/applications/common/schemaform/fields/ArrayField.jsx
Comment (cehsu) [reply src/applications/common/schemaform/fields/ArrayField.jsx]: Good point, @jbalboni, a helper function `getItemModes` has been set up and is now used here instead.

**Comment by cehsu** [comment]
@annekainicUSDS please let us know when this can be merged in/the forms library code freeze ends.

**Review by markgreenburg — CHANGES_REQUESTED** [review]
Been a while since I've taken a look at this :)
My main concern is that there is no 'remove' button for new service periods, like there is for pre-filled service periods. This is a problem because if a user hits 'add new service period', they will not be able to continue with the 526 until the enter in some fake 'required' information. Happy to re-review once you're able to figure that one out!

**Comment by cehsu** [comment]
We could remove the condition in front of the remove button @markgreenburg, is that what you recommend? It doesn't seem like this would require design approval?

**Comment by cehsu** [comment]
@jbalboni @goldenmeanie is there any reason we don't provide a remove button for new array items? If not, we can remove the conditional check and always provide a remove button, as recommended by @markgreenburg.

**Comment by jbalboni** [comment]
I would need to understand the page more, but typically we've tried not to have pages with array fields that have a minItems of 0 without putting a yes/no question in to hide the field. If you're looking to change the default behavior for array fields I'd want to see your use case, because I think adding that button will cause problems on other forms.

**Comment by cehsu** [comment]
Good point @jbalboni, if I understand you correctly, we could use a condition to check whether the minItems (typically 1) requirement has been reached, if not, don't add the button.

**Comment by jbalboni** [comment]
Looking at it again, I think what you want to do is show the remove button when you're adding and `reviewMode` is true. I think that will keep things working as expected on the other forms.

File: src/applications/common/schemaform/fields/ArrayField.jsx
Comment (cehsu) [inline src/applications/common/schemaform/fields/ArrayField.jsx]: @markgreenburg actually we do already provide the remove buttons to added items (but we did not previously provide the update button to the last item -- which needed to be updated as we now allow last items to be in edit mode), I've fixed the bug here. Nice catch!

**Comment by cehsu** [comment]
Closing this and will open a new PR on the new us-forms-system repo.

**Pull request closed without merging**
